lib/CSSStyleDeclaration.js

```diff
@@ -162,133 +99,78 @@ class CSSStyleDeclaration {
+    this._checkReadonly();
+    if (this._updating) {
       return;
     }
-    this._setInProgress = true;
+    this._clearProperties();
+    this._updating = true;
     try {
-      const valueObj = parseCSS(
-        val,
-        {
-          context: "declarationList",
-          parseValue: false
-        },
-        true
-      );
+      const valueObj = this._parseCSSText(val);
       if (valueObj?.children) {
-        const properties = new Map();
-        let shouldSkipNext = false;
-        for (const item of valueObj.children) {
-          if (item.type === "Atrule") {
-            continue;
-          }
-          if (item.type === "Rule") {
-            shouldSkipNext = true;
-            continue;
-          }
-          if (shouldSkipNext === true) {
-            shouldSkipNext = false;
-            continue;
-          }
-          const {
-            important,
-            property,
-            value: { value }
-          } = item;
-          if (typeof property === "string" && typeof value === "string") {
-            const priority = important ? "important" : "";
-            const isCustomProperty = property.startsWith("--");
-            if (isCustomProperty || hasVarFunc(value)) {
-              if (properties.has(property)) {
-                const { priority: itemPriority } = properties.get(property);
-                if (!itemPriority) {
-                  properties.set(property, { property, value, priority });
-                }
-              } else {
-                properties.set(property, { property, value, priority });
-              }
-            } else {
-              const parsedValue = parsePropertyValue(property, value, {
-                globalObject: this._global
-              });
-              if (parsedValue) {
-                if (properties.has(property)) {
-                  const { priority: itemPriority } = properties.get(property);
-                  if (!itemPriority) {
-                    properties.set(property, { property, value, priority });
-                  }
-                } else {
-                  properties.set(property, { property, value, priority });
-                }
-              } else {
-                this.removeProperty(property);
-              }
-            }
-          }
-        }
-        const parsedProperties = prepareProperties(properties, {
-          globalObject: this._global
-        });
-        for (const [property, item] of parsedProperties) {
-          const { priority, value } = item;
-          this._priorities.set(property, priority);
-          this.setProperty(property, value, priority);
-        }
+        const properties = this._extractDeclarations(valueObj.children);
+        this._applyProperties(properties);
       }
     } catch {
       return;
     }
-    this._setInProgress = false;
-    if (typeof this._onChange === "function") {
-      this._onChange(this.cssText);
-    }
+    this._updating = false;
```

Comment:
Putting this outside the catch block seems likely to lead to bugs. I noticed this pattern, of setting a boolean at the top and resetting it outside, in some of the individual files to.

There are two good solutions:

- Use a `finally` block
- Invent a custom disposable and use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/using, so that when the scope is ended it always gets fixed. E.g. something like `using updater = this._startUpdate();` at the top of the function.